Add AGLight Export Plugin

hrueger opened the pull request:
Hi,
you might remember issue #1226 .
As we discussed there, I wanted to create an export plugin to use the fixture data of this excellent library in AGLight.

Please do not merge this as of now, as it is still work in progress.

hrueger commented:
According to [the docs](https://github.com/OpenLightingProject/open-fixture-library/blob/master/docs/ui.md#deployment), this should be available at https://open-fixture-library-pr-1357.herokuapp.com/ , right?

FloEdelmann commented:
Thanks @hrueger for this PR! I will have a deeper look at it in a few days.

> this should be available at open-fixture-library-pr-1357.herokuapp.com

It has to be triggered manually for PRs from non-contributors; I've done that now. The review app is now available at <https://open-fixture-library-pr-1357.herokuapp.com/>. Note that it expires after 5 days without a new commit, so just ping me if you need it again.

FloEdelmann commented on plugins/aglight/colors.js:
```diff
@@ -0,0 +1,605 @@
+/** Taken from https://github.com/timoxley/colornames */
```
We already use the [color-names](https://github.com/jonathantneal/color-names) npm module in the e:cue plugin: https://github.com/OpenLightingProject/open-fixture-library/blob/schema-12.2.1/plugins/ecue/import.js#L1-L10

Is there a way to use that module here, too?
Or maybe both plugins can be refactored to use another npm module? I'd rather not have a list of color names to maintain myself.

hrueger replied:
I found that module, but I didn't use it as it is 5+ years old... Anyway, the color names haven't changed much 😉
I'll change that 👍

hrueger replied:
What about using this?
[color-name-list](https://github.com/meodai/color-names/) (or on [npm](https://www.npmjs.com/package/color-name-list))

It is maintained and has a lot more colors.

I'm missing the following colors in the color-names npm module: `UV`, `Cold white` and `Warm White`. Of course, I could just add them in the code of the export plugin, but I don't feel like this would be good practise...

FloEdelmann replied:
That module looks good! :)
Would you mind changing the e:cue plugin to that module in a separate PR, too?

hrueger replied:
No problem, I can do that. Thanks for re-deploying

FloEdelmann commented on plugins/aglight/export.js:
```diff
@@ -0,0 +1,168 @@
+/* eslint-disable radix */
```
Why did you disable this rule? It can be easily silenced by passing the second (radix) parameter to `parseInt` to avoid nasty unexpected bugs: https://eslint.org/docs/rules/radix

```
parseInt(someString, 10)
```

FloEdelmann commented on plugins/aglight/export.js:
```diff
@@ -0,0 +1,168 @@
+/* Based on the ofl export plugin */
+
+const fixtureJsonStringify = require(`../../lib/fixture-json-stringify.js`);
+const colors = require(`./colors`);
```
Please include the file extension.

```suggestion
const colors = require(`./colors.js`);
```

FloEdelmann commented on plugins/aglight/export.js:
```diff
@@ -0,0 +1,168 @@
+ * @returns {Promise.<Array.<Object>, Error>} The generated files.
+ */
+module.exports.export = async function exportAGLight(fixtures, options) {
+  console.log(options);
```
Debug log

FloEdelmann commented on plugins/aglight/export.js:
```diff
@@ -0,0 +1,168 @@
+    version: displayedPluginVersion,
+    fixtures: [],
+  };
+    // one JSON file for each fixture
```
This comment is no longer valid, as you only return one large JSON file at the end of the function.

FloEdelmann commented on plugins/aglight/export.js:
```diff
@@ -0,0 +1,168 @@
+
+/**
+ * @param {Object} content The fixture data
+ * @returns {Object} The transformed fixture
```
No need to return the same object again; the editing happens in place (see https://stackoverflow.com/a/6605700/451391).

FloEdelmann commented on plugins/aglight/export.js:
```diff
@@ -0,0 +1,168 @@
+ */
+function transformSingleCapabilityToArray(content) {
+  if (content.availableChannels) {
+    for (const k of Object.keys(content.availableChannels)) {
+      if (content.availableChannels[k].capability) {
+        content.availableChannels[k].capabilities = [content.availableChannels[k].capability];
+        content.availableChannels[k].singleCapability = true;
+        delete content.availableChannels[k].capability;
+      }
+    }
```
You can use `Object.values` to make the following lines simpler:

```suggestion
    for (const channel of Object.values(content.availableChannels)) {
      if (channel.capability) {
        channel.capabilities = [channel.capability];
        channel.singleCapability = true;
        delete channel.capability;
      }
    }
```

FloEdelmann commented on plugins/aglight/export.js:
```diff
@@ -0,0 +1,168 @@
+    fixtures: [],
+  };
+    // one JSON file for each fixture
+  library.fixtures = fixtures.map(fixture => {
```
Please put the assignment directly where the `library` object is created, i.e.

```js
const library = {
  version: displayedPluginVersion,
  fixtures: fixtures.map(fixture => {
    // …
  }),
};
```

FloEdelmann commented on plugins/aglight/export.js:
```diff
@@ -0,0 +1,168 @@
+ * @param {Object} content The fixture data
+ * @returns {Object} The transformed fixture
+ */
+// eslint-disable-next-line complexity
```
I know just disabling the rule is tempting, but we enabled it for a reason. Could you maybe split off parts of your function into separate functions with speaking names, so that reading and understanding the function gets easier?

hrueger commented:
@FloEdelmann thank's a lot!

I also added two new fixtures, #1363 and #1364 .
So when you have time, please check them first. This PR is not as important as it is only me and a few friends using the AGLight software as of now. But we only have a few different fixtures for testing. That's why I added those two fixtures. So those fixtures help us more than this PR (we can also manually copy the library file from our local installation).

Thank's in advance.

FloEdelmann commented on plugins/aglight/plugin.json:
```diff
@@ -0,0 +1,31 @@
+    "$schema": "../../schemas/plugin.json",
+
+    "name": "AGLight v1.0.0",
+    "previousVersions": {},
```
You can remove this line

hrueger replied:
The `previousVersions` or the blank line?

FloEdelmann replied:
`previousVersions` is not needed when it's empty.

hrueger commented:
@FloEdelmann Could you please redeploy? I'm one day too late ;-)

hrueger commented:
I think I'm finished here.
GitHub now tells me that there still are some changes requested, however, all conversations are marked as resolved. Did I miss something which needs to be done?

FloEdelmann commented on plugins/aglight/plugin.json:
```diff
@@ -0,0 +1,30 @@
+{
+  "$schema": "../../schemas/plugin.json",
+
+  "name": "AGLight v1.0.0",
```
One more thing: A version string in the plugin name refers to the supported file format version, not the plugin version. I guess you can just drop it in this case. See e.g. https://open-fixture-library.org/about/plugins/qlcplus_4.12.2 how the QLC+ format version, the import plugin version and the export plugin version differ.

```suggestion
  "name": "AGLight",
```

Note that you also need to run `make` to keep `plugins.json` updated.

hrueger commented:
>Could you please have a look if it works fine for you?

Sure! However, decimal places are now truncated... Is this because we're using the model now? Anyway, it's not a huge problem.
![grafik](https://user-images.githubusercontent.com/18656830/88528605-243fea00-cfff-11ea-874d-7dabfb96d318.png)

>Do you really need the separate resources? All gobo resources used in the downloaded fixtures are included in the fixture JSON.

Oh, of course not. I didn't recognize that...

> Could you please improve the descriptions in the JSDoc comments?

I'll try and do my best.

FloEdelmann commented:
> However, decimal places are now truncated... Is this because we're using the model now?

I think it's the other way around; your original code used `parseInt`, while in the Entity class, we use `parseFloat`:

https://github.com/OpenLightingProject/open-fixture-library/blob/5ba24eec91ebc4126f363dcea112990e95248bda/lib/model/Entity.js#L152

Thanks for the updated commits, I will have a final look this evening :)

hrueger commented:
>I think it's the other way around

Oops, so I compared the files the wrong way round. Sorry for that ;-)

FloEdelmann commented:
@hrueger I've made another few changes; can you again please check if they look okay?

Two changes could be breaking on your side:

1. I've removed the `templateChannels`: [`43fba87`](https://github.com/OpenLightingProject/open-fixture-library/pull/1357/commits/43fba87c0656cad9908bda96e664deec7b64ade5)
2. Resolved matrix channels don't contain the `matrixChannel` property anymore, just the `matrixChannelKey` (pixel key): [`27f9873`](https://github.com/OpenLightingProject/open-fixture-library/pull/1357/commits/27f9873355eb9af34a8b923e58b8dad1bb369c9c)
  That's a tradeoff we have to make if we want to leverage the model here, which fixes a lot of issues with your original implementation (e.g. wrong `matrixChannelKey`s, missing matrix channels when using a string in `repeatFor`).

When testing the changes, you can use the [diff-plugin-outputs.js script](https://github.com/OpenLightingProject/open-fixture-library/blob/master/cli/diff-plugin-outputs.js) to track changes in the output of the plugin between the currently checked out version and another git ref:

```
./cli/diff-plugin-outputs.js --help
./cli/diff-plugin-outputs.js --plugin aglight --ref 629e996 fixtures/man/key.json
```

hrueger commented:
Thanks, I'll have a look at it tomorrow

hrueger commented:
Sorry for the delay.

Yes, I'm fine with that. In the current version of AGLight, template channels are treated as normal channels and I planned using `matrixChannel` and `matrixChannelKey` in a future update. So I didn't have to change anything.

Thank's for your work!

FloEdelmann commented:
> I planned using `matrixChannel` and `matrixChannelKey` in a future update

Thanks for your answer. Note that `matrixChannel` (e.g. `Red $pixelKey`) is not included by the current version, just `matrixChannelKey` (aka `pixelKey`, e.g. `Red 1`).

If you aren't using the key anyway as of now, would you mind if we rename it to `pixelKey` to make the naming more consistent?

hrueger commented:
Sure

hrueger commented:
Thank's a lot! 😀
